feat: initial overscroll background color and iOS bounce implementation from msdewitt

trevinhofmann opened the pull request:
This PR fixes #1700 by setting the overscroll background color from the theme when relaunching the Capacitor app, and includes a fix to keep the iOS bounce.

This PR has been built upon #2594, rebasing and squashing to resolve conflicts and clean up the commit history.

- [X] Squash to remove large `node_modules` updates from the commit history
- [x] Ensure that the webview package uses the same TS version as the main app (https://github.com/cybersemics/em/pull/2594#issuecomment-2561219316)
- [x] Remove unused files from webview bundle config (https://github.com/cybersemics/em/pull/2594#discussion_r1902208184)
- [x] Switch the webview package to yarn (https://github.com/cybersemics/em/pull/2594#discussion_r1902210622)
- [x] Remove irrelevant changes (https://github.com/cybersemics/em/pull/2594#discussion_r1902211403)
- [x] Build the webview package from source (https://github.com/cybersemics/em/pull/2594#discussion_r1902209905)
- [x] Remove unintentional comment (https://github.com/cybersemics/em/pull/2594#discussion_r1902217858)
- [x] Fix the positioning of the toolbar (https://github.com/cybersemics/em/pull/2594#issuecomment-2589184873)

raineorshine commented on src/components/AppComponent.tsx:
```diff
@@ -125,6 +126,7 @@ const AppComponent: FC = () => {
   const tutorial = useSelector(isTutorial)
   const splitPosition = useSelector(state => state.splitPosition)
 
+  WebviewBackground.changeBackgroundColor({ color: colors.bg })
```
Should we put this in a `useEffect`? We usually want to avoid side effects directly in component render functions.

trevinhofmann replied:
Good call. Placed inside a `useEffect` here: 25268172daa0c7e55f005a587e9c803a5f2293d3

raineorshine commented on src/components/ThemeSwitch.tsx:
```diff
@@ -1,13 +1,17 @@
   const dispatch = useDispatch()
   const light = useSelector(state => theme(state) === 'Light')
+  const colors = useSelector(themeColors)
+  WebviewBackground.changeBackgroundColor({ color: colors.bg })
```
Similarly, we might want to put this into a `useEffect` hook so that it is only called when the color changes.

trevinhofmann replied:
Placed inside a `useEffect` hook also: 25268172daa0c7e55f005a587e9c803a5f2293d3

raineorshine commented on src/components/modals/ModalComponent.tsx:
```diff
@@ -66,9 +67,12 @@ class ModalComponent extends React.Component<ModalProps> {
+  close = (fullReload: any = false) => {
     this.animateAndClose!()
     this.props.onClose?.()
+    if (Capacitor.isNativePlatform() && fullReload) {
+      window.location.reload()
+    }
```
`fullReload` can be removed if reloading the page is no longer necessary.

trevinhofmann replied:
It seems we _can_ remove this. I need to look into why it's no longer necessary, but it seems to be working fine without the reload.

Removed: 4792aeb5696e436ff39808f46a8e7ecd6ad16d86

raineorshine reviewed (changes requested):
Thank you! 

I have tested this on a physical iPhone and everything works as expected. The overscroll background matches the theme when the setting is changed, and it persists when the app is reloaded. iOS overscroll bounce works and no other changes in behavior are observed.

Great job setting up the build steps for this.

---

I was [previously](https://github.com/cybersemics/em/pull/2594#discussion_r1859389869) under the impression that it was not possible to change the background color of the WebView without reloading the page. If I am not mistake, that appears to no longer be the case.  (At least, the background color is correct when I overscroll on the Settings screen, before any reload occurs.) Could you briefly explain what made this possible?

---

When I run `yarn cap:sync` it changes some files under version control. I would expect a build step to not modify files under version control. Are you seeing the same thing?

```diff
diff --git a/android/app/capacitor.build.gradle b/android/app/capacitor.build.gradle
index 44eaa5c5383..b5234abf6f0 100644
--- a/android/app/capacitor.build.gradle
+++ b/android/app/capacitor.build.gradle
@@ -13,7 +13,6 @@ dependencies {
     implementation project(':capacitor-haptics')
     implementation project(':capacitor-keyboard')
     implementation project(':capacitor-status-bar')
-    implementation project(':webview-background')
 
 }
 
diff --git a/android/capacitor.settings.gradle b/android/capacitor.settings.gradle
index 2f308cf70ba..be4207a2625 100644
--- a/android/capacitor.settings.gradle
+++ b/android/capacitor.settings.gradle
@@ -13,6 +13,3 @@ project(':capacitor-keyboard').projectDir = new File('../node_modules/@capacitor
 
 include ':capacitor-status-bar'
 project(':capacitor-status-bar').projectDir = new File('../node_modules/@capacitor/status-bar/android')
-
-include ':webview-background'
-project(':webview-background').projectDir = new File('../node_modules/webview-background/android')
diff --git a/ios/App/Podfile b/ios/App/Podfile
index fe1eddeba9b..9235d198149 100644
--- a/ios/App/Podfile
+++ b/ios/App/Podfile
@@ -15,7 +15,7 @@ def capacitor_pods
   pod 'CapacitorHaptics', :path => '../../node_modules/@capacitor/haptics'
   pod 'CapacitorKeyboard', :path => '../../node_modules/@capacitor/keyboard'
   pod 'CapacitorStatusBar', :path => '../../node_modules/@capacitor/status-bar'
-  pod 'WebviewBackground', :path => '../../node_modules/webview-background'
+  pod 'WebviewBackground', :path => '../../packages/webview'
 end
 
 target 'App' do
```

trevinhofmann commented on packages/webview/.swiftpm/xcode/xcuserdata/matheudewitt.xcuserdatad/xcschemes/xcschememanagement.plist:
Good catch. Yes, removed: 0c9220985125274229bfa200bf743e0f7b32b1e2

trevinhofmann commented:
> Great job setting up the build steps for this.

Thanks! Took a bit to figure out yarn workspaces, but it seems to be the best approach for building from source and keeping everything in sync.

> I was https://github.com/cybersemics/em/pull/2594#discussion_r1859389869 under the impression that it was not possible to change the background color of the WebView without reloading the page. If I am not mistake, that appears to no longer be the case. (At least, the background color is correct when I overscroll on the Settings screen, before any reload occurs.) Could you briefly explain what made this possible?

I removed this reload and noticed we were still having issues on mobile (both iOS Safari and Android Chrome). After looking into it, we just need to also update the `<html>` background-color and the meta `theme-color` to switch the overscroll color and tab/theme color for these platforms.

I'm not _entirely_ familiar with the original reason we may have added the page reload, but it likely fixed this issue on some platforms since the `index.html` script updates these values from the theme during the initial page load.

By adding this change (which also removed the `useEffect` in `ThemeSwitch` as I've realized it seems to be redundant with the one in `AppComponent`), everything seems to be working correctly for me on all platforms with the page reload: e462d6114bee23f5e6861d0f961b8cd6501e5e55

> When I run yarn cap:sync it changes some files under version control. I would expect a build step to not modify files under version control. Are you seeing the same thing?

Yep! Looks like the committed changes were missing some config updates. Updated here: 63f741b751a6433f759abe09f32345106823ba3f

raineorshine reviewed (approved):
Amazing. The extra effort here really paid off with both better UX and cleaner code. Well done.
